### PR TITLE
Adding parsing of threshold_backward and _softmax for LTC

### LINKED ISSUE #1309: Add `aten::_softmax` parsing
## 🚀 Feature
Need to add parser support in `parser.cpp` for `aten::_softmax` op.  LTC traces the `_softmax` op variant that is different from TorchScript.

### PR DESCRIPTION
**Pull request by kevinstephano**
Note: symbolic_script.py was edited to test threshold_backward by inserting it into the autodiff of relu but this may not be appropriate to upstream.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 1]
```diff
@@ -1052,6 +1052,37 @@ class IrParser {
...
 
+    { // LTC uses threshold_backward for relu_backward
+      auto ptr_op = getOperatorForLiteral(
+          "aten::threshold_backward(Tensor grad_output, Tensor self, Scalar threshold) -> Tensor");
+      REGISTER_PARSE_RULE(
+          ptr_op,
+          {
+            MemoryFormat format;
+            std::list<Val*> list_val;
+            std::tie(format, list_val) = getConsistentValues(
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/parser.cpp:1063]: nitpick: we should be able to easily handle permutation for pw operations.
e.g.
https://github.com/csarofeen/pytorch/blob/ddaa2a5535c5f44af5f8c21682a41489e149e1d7/torch/csrc/jit/codegen/cuda/parser.cpp#L710-L713
https://github.com/csarofeen/pytorch/blob/ddaa2a5535c5f44af5f8c21682a41489e149e1d7/torch/csrc/jit/codegen/cuda/parser.cpp#L727-L728

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 2]
```diff
@@ -1833,6 +1864,56 @@ class IrParser {
...
+            auto dim_value = constant_as<int>(node->input(1));
+            TORCH_INTERNAL_ASSERT(
+                dim_value.has_value(), "dim in softmax is not valid");
+
+            auto output = softmax(input, dim_value.value());
+            value_map.emplace(node->output()->unique(), output);
+          },
+          [](const Node* node) -> bool {
+            if (node->inputs()[1]->node()->kind() != prim::Constant) {
+              return false;
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/parser.cpp:1891]: Nice!

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 3]
```diff
@@ -1833,6 +1864,56 @@ class IrParser {
...
+            }
+            if (node->inputs()[2]->node()->kind() == prim::Constant) {
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/parser.cpp]: nitpitck: could simplify the logic down a little bit.

```
const auto half_to_float = constant_as<bool>(node->input(2));
auto input_tensor_type = node->input(0)->type()->cast<TensorType>();
// we don't support `half_to_float == false` or `half_to_float == true && input.dtype != half`
if (half_to_float.has_value() && (!half_to_float.value() || input_tensor_type->scalarType() != at::ScalarType::Half)) {
  return false;
}
return true; 
```

**Review by jjsjann123 — COMMENTED** [review]
Minor comments

File: torch/csrc/jit/codegen/cuda/parser.cpp (hunk 1 above)
Comment (kevinstephano) [reply torch/csrc/jit/codegen/cuda/parser.cpp:1063]: Fixed.

File: torch/csrc/jit/codegen/cuda/parser.cpp (hunk 3 above)
Comment (kevinstephano) [reply torch/csrc/jit/codegen/cuda/parser.cpp]: I significantly fixed up this logic to make it simpler. I think there was a slight bug in your suggestion as we do want to accept `False` from the boolean.

Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/parser.cpp]: Oops, sorry about my buggy suggestion. :stuck_out_tongue:

**Review by jjsjann123 — APPROVED** [review]
LGTM

**Pull request merged**